File: 2024/day-2/day-2.js
const fs = require('fs')

const allReports = []

fs.readFileSync('ip.txt', 'utf8')
    .split('\n')
    .map(report => allReports.push(report.trim().split(' ')))

    console.log('all reports', allReports.length)

// a report only counts as safe if both of the following are true:
// 1. The levels are either all increasing or all decreasing.
// 2. Any two adjacent levels differ by at least one and at most three.
function isSafe(report) {
    if (report.length <= 1) return true;
    
    let isIncreasing = true;
    let isDecreasing = true;
    
    for (let i = 1; i < report.length; i++) {
        const diff = report[i] - report[i - 1];
        
        if (Math.abs(diff) < 1 || Math.abs(diff) > 3) {
            return false;
        }
        
        if (diff < 0) {
            isIncreasing = false;
        } else if (diff > 0) {
            isDecreasing = false;
        }
        
        if (!isIncreasing && !isDecreasing) {
            return false;
        }
    }
    
    return true;
}

function canBeSafe(report) {
    // check if the array can be safe after removing one element
    for (let i = 0; i < report.length; i++) {
        let newReport = report.slice(0, i).concat(report.slice(i + 1))
        if (isSafe(newReport)) {
            return true
        }
    }
    return false
}

let cnt = 0
for (let i = 0; i < allReports.length; i++) {
    if (isSafe(allReports[i])) {
        cnt += 1
    } else if (canBeSafe(allReports[i])) {
        cnt += 1
    }
}

console.log('cnt', cnt)

File: 2024/day-2/day-2.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;
vector<vector<int> > data;

void loadData() {
  string line;
  ifstream ip("ip.txt");
  if (ip.is_open()) {

    while (getline(ip, line)) {
        string level;
        vector<int> report;
        stringstream ss(line);
        while (getline(ss, level, ' ')) {
            report.push_back(stoi(level));
        }

        data.push_back(report);
    }

    ip.close();
  }
}

// part-1: a report only counts as safe if both of the following are true:
// 1. The levels are either all increasing or all decreasing
// 2. Any two adjacent levels differ by at least one and at most three
bool isSafe(vector<int> arr) {
    int n = arr.size();
    bool isIncreasing = true;
    bool isDecreasing = true;

    for (int i = 0; i < n - 1; i++) {
        if (arr[i] >= arr[i + 1]) {
            isIncreasing = false;
        }
        if (arr[i] <= arr[i + 1]) {
            isDecreasing = false;
        }
        if (abs(arr[i] - arr[i + 1]) > 3) {
            return false;
        }
    }

    return isIncreasing || isDecreasing;
}

// part-2: Now, the same rules apply as before, except if removing a single level from an unsafe report would make it safe, the report instead counts as safe.
bool canBeSafe(vector<int> arr) {
    int n = arr.size();
    for (int i = 0; i < n; i++) {
        vector<int> temp = arr;
        temp.erase(temp.begin() + i);
        if (isSafe(temp)) {
            return true;
        }
    }

    return false;
}

int main() {
  loadData();
  int cnt = 0;

  for (int i = 0; i < data.size(); i++) {
    if (isSafe(data[i])) { // part - 1
      cnt++;
    } else if (canBeSafe(data[i])) { // part - 2
      cnt++;
    }
  }

  cout << "total safe reports: " << cnt;

  return 0;
}